editor/src/viewport_tools/tools/spline.rs

```diff
@@ -0,0 +1,270 @@
+
+					data.weight = tool_options.line_weight;
+
+					responses.push_back(make_operation(data, tool_data, true));
+
+					Drawing
+				}
+				(Drawing, DragStop) => {
+					let snapped_position = data.snap_handler.snap_position(responses, input.viewport_bounds.size(), document, input.mouse.position);
+					let pos = transform.inverse().transform_point2(snapped_position);
+
+					// TODO: introduce comparison threshold when operating with canvas coordinates (https://github.com/GraphiteEditor/Graphite/issues/100)
+					if data.points.last() != Some(&pos) {
+						data.points.push(pos);
+						data.next_point = pos;
+					}
+
+					responses.push_back(remove_preview(data));
+					responses.push_back(make_operation(data, tool_data, true));
+
+					Drawing
+				}
+				(Drawing, PointerMove) => {
+					let snapped_position = data.snap_handler.snap_position(responses, input.viewport_bounds.size(), document, input.mouse.position);
+					let pos = transform.inverse().transform_point2(snapped_position);
+					data.next_point = pos;
+
+					responses.push_back(remove_preview(data));
+					responses.push_back(make_operation(data, tool_data, true));
+
+					Drawing
+				}
+				(Drawing, Confirm) | (Drawing, Abort) => {
+					if data.points.len() >= 2 {
+						responses.push_back(DocumentMessage::DeselectAllLayers.into());
+						responses.push_back(remove_preview(data));
+						responses.push_back(make_operation(data, tool_data, false));
+						responses.push_back(DocumentMessage::CommitTransaction.into());
+					} else {
+						responses.push_back(DocumentMessage::AbortTransaction.into());
+					}
+
+					data.path = None;
+					data.points.clear();
+					data.snap_handler.cleanup(responses);
+
+					Ready
+				}
+				_ => self,
+			}
+		} else {
+			self
+		}
+	}
+
+	fn update_hints(&self, responses: &mut VecDeque<Message>) {
+		let hint_data = match self {
+			SplineToolFsmState::Ready => HintData(vec![HintGroup(vec![HintInfo {
+				key_groups: vec![],
+				mouse: Some(MouseMotion::Lmb),
+				label: String::from("Draw Spline"),
+				plus: false,
+			}])]),
+			SplineToolFsmState::Drawing => HintData(vec![
+				HintGroup(vec![HintInfo {
+					key_groups: vec![],
+					mouse: Some(MouseMotion::Lmb),
+					label: String::from("Extend Spline"),
+					plus: false,
+				}]),
+				HintGroup(vec![HintInfo {
+					key_groups: vec![KeysGroup(vec![Key::KeyEnter])],
+					mouse: None,
+					label: String::from("End Spline"),
+					plus: false,
+				}]),
+			]),
+		};
+
+		responses.push_back(FrontendMessage::UpdateInputHints { hint_data }.into());
+	}
+
+	fn update_cursor(&self, responses: &mut VecDeque<Message>) {
+		responses.push_back(FrontendMessage::UpdateMouseCursor { cursor: MouseCursorIcon::Default }.into());
+	}
+}
+
+fn remove_preview(data: &SplineToolData) -> Message {
+	Operation::DeleteLayer { path: data.path.clone().unwrap() }.into()
+}
+
+fn make_operation(data: &SplineToolData, tool_data: &DocumentToolData, show_preview: bool) -> Message {
```

Comment:
more descriptive name than `make_operation`? something like `add_spline` or something like that?

Reply:
Done!